### ISSUE TITLE
How to use the API model "MultiTaskGP" or others to fit some specific data?

### ISSUE THREAD

**Issue by githubwl2021** [pytorch/botorch#1674]
## Issue description

We are facing some issues for using the API model "MultiTaskGP", which is a Hadamard multi-task, multi-output GP. First, let me introduce our use case. Suppose we have 2 inputs (x1, x2). Also, we have 2 tasks (t1, t2), one of which has 9 outputs (o1,...,o9) while the other one has 4 outputs (o10,...,o13). Our case is that we have all output observations for the 1st task t1, but for the 2nd task t2, we only have partial output observations. Please let me illustrate our data further as below.

```
x1	x2	o1, 	o2,	o3,	o4,	o5,	o6,	o7,	o8,	o9,	o10, 	o11, 	o12, 	o13
--------------------------------------------------------------------------------------------------------------------
1	2	1	2	3	4	5	6	7	8	9	NA	 NA   	NA   	NA
...

11	22	11	22	33	44	55	66	77	88	99	10	11	12	13
...
```
From the above data illustration, we can see that our data can be divided into 2 categories: 1) with only task 1 data from o1 to o9; 2) with both task 1 and task 2 data from o1 to o13. Can we build one GP model using the API model "MultiTaskGP" to fit the above 2 categories of data simultaneously?

Btw, if we treat each output as an individual task, we will have 13 tasks (t1, t2, ..., t13). Each task has 1 output. However, the partial observation issue still exists. To solve this problem, can anyone give any advice or comment? Thanks.

## Code example

Below is the example provided with the API model"MultiTaskGP" in the botorch tutorial.
```
>>> X1, X2 = torch.rand(10, 2), torch.rand(20, 2)
>>> i1, i2 = torch.zeros(10, 1), torch.ones(20, 1)
>>> train_X = torch.cat([torch.cat([X1, i1], -1), torch.cat([X2, i2], -1),])
>>> train_Y = torch.cat(f1(X1), f2(X2)).unsqueeze(-1)
>>> model = MultiTaskGP(train_X, train_Y, task_feature=-1)
```
This example is with limited information for us to transfer to our case. So, we also want to ask for the exact meaning of the arguments of "task_feature" and "output_tasks", and how to use them in our application. Thanks.

**Comment by Balandat** [comment]
So the `MultiTaskGP` is a relatively simple model that assumes each "task" corresponds to one observed output. `MultiTaskGP` by default fits a standard ICM kernel across all these "tasks". The "output_tasks" setting is a convenience handler so that you can have  model that during prediction doesn't predict all tasks. 

In your case, you have outputs grouped by tasks, and so this is a richer structure (a multi-task multi-output model). You could define some other multi-task kernel structure for this (e.g. one where outputs are correlated within a task, but across tasks only through a common task-level correlation parameter). 

You can combine things by modeling each output as an individual task. The standard `MultiTaskGP` will be able to model the correlations between the different outputs but it will not allow to specify such richer structure. Also, estimating this rather large cross-task-covariance matrix will require a lot of data, and you'll likely run into scalability issues sooner rather than later with this model. 

That said, if you model things individually then there is no "partial observation issue" - each data point will just be `((x1, x2, i), y)`, where `i` is the task index and `y` the associated observation. If you have two observations for different tasks at the same parameters `x1, x2`, you just pass that in as another data point.

**Comment by githubwl2021** [comment]
@Balandat Thank you very much for your quick reply.

Please let me ask some further questions as follows.

If we use `MultiTaskGP` to model each output as an individual task, how should we set the argument `task_feature`? The range of `task_feature` is `(-d, d)`, where `d` is the input dimension. What is the principle behind this range? Why this range is related to the input dimension? How does this argument `task_feature` affect the GP model?

In the tutorial example regarding the API model `MultiTaskGP`, the argument `task_feature` is set to `-1`.  What does this `-1` mean in this example, where `d=2` and the number of tasks is also `2`. However, for our case, the input dimension is 2, while the number of total tasks is 13. How should we set this argument `task_feature` if we want to model all data for 13 tasks together? Can you please advice on this further? Thanks.

**Comment by Balandat** [comment]
One of the features in the training data to the model is the task feature. If `p` is the number of features you're modeling over (your `x`), then the model expects a `d = p+1` sized training data, where the `task_feature` element indicates the column index for the task indicator. 

In your case `p=2`, the input is the concatenated observations `x` together with the task indicator, also three-dimensional in your case. It just means that if you have 13 observations for the same `x` you need to pass in 13 rows of data, each with a different task indicator.

**Comment by githubwl2021** [comment]
@Balandat Thanks a lot. I see. So if we also put the task indicator to the last column as in the tutorial example, it means that we should also use `task_feature = -1`. Is that correct? Thanks.

**Comment by Balandat** [comment]
Correct!

**Comment by githubwl2021** [comment]
Hi Balandat, 

Sorry to disrupt you again.

If one of our task/output value is negative, the trained Multi-Task Gaussian Process (MTGP) model should give predictions with negative values for this task/output, right? However, our GP model sometime give positive predictions for this negative task, and sometime give negative predictions for this negative task. Even, the model could give negative predictions for the positive tasks. Do you know usually what cause this uncertainty for predictions? According to our experience, the Multi-Objective Gaussian Process (MOGP) model does not have this issue.

Another quick question is regarding the task index. As we have 13 tasks, is that we just use 0, 1, 2, ..., 12 for these 13 task indices? Thanks.

**Comment by githubwl2021** [comment]
As mentioned above, we are trying to use the `ConstrainedMCObjective` to constrain our outputs. Below is the example behind this API model. 

```
>>> bound = 0.0
>>> objective = lambda Y: Y[..., 0]
>>> # apply non-negativity constraint on f(x)[1]
>>> constraint = lambda Y: bound - Y[..., 1]
>>> constrained_objective = ConstrainedMCObjective(objective, [constraint])
>>> samples = sampler(posterior)
>>> objective = constrained_objective(samples)
```

Regarding this example, we have several questions as follows.

1) What is the meaning of `objective = lambda Y: Y[..., 0]` here?
2) We use the constraint as `constraint = lambda Y: bound - Y[..., 1]` in our codes. But, the output predictions still have negative values. Why?
3) As mentioned before, we actually want some outputs larger than 0, but some of them smaller than 0. For our case, how should we set constraints for different outputs (13 tasks/outputs)?

Thanks a lot in advanced!

**Comment by saitcakmak** [comment]
Hi @githubwl2021. Sorry for the late reply.

> What is the meaning of objective = lambda Y: Y[..., 0] here?

That says to use the first outcome (`Y[..., 0]`) as the objective value.

> We use the constraint as constraint = lambda Y: bound - Y[..., 1] in our codes. But, the output predictions still have negative values. Why?

I am not sure what the "predictions" is referring to here. If you're referring to the samples from the model posterior, the objective is applied after the samples are generated, so those would not be affected by the constraint. For the objective values themselves, the application of the constraints is a bit involved. 
- There's an `infeasible_cost` input to `ConstrainedMCObjective`. The goal of `infeasible_cost` is to ensure that `objectives + infeasible_cost` is non-negative, which is required for proper application of the constraints. The constraints are applied to `objectives + infeasible_cost`, then the result is brought back to the original scale by subtracting the `infeasible_cost`. This results in the infeasible objective values being `infeasible_cost` after the constraints are applied.
- We apply "soft" constraints using a sigmoid transform to ensure differentiability of the objective values. You can control how soft this is with the `eta` parameter. This means that if the constraint value is 0, the sigmoid would evaluate to 0.5, so that the objective value would be half way between the raw objective (`Y[..., 0]`) and the infeasible cost. As the constraint value increases (i.e., the sample becomes more infeasible), the objective value will move closer to the `infeasible_cost` and vice-versa. 

> As mentioned before, we actually want some outputs larger than 0, but some of them smaller than 0. For our case, how should we set constraints for different outputs (13 tasks/outputs)?

Hopefully the above points help explain how things work under the hood. To handle multiple constraints, you can pass in a list of constraints to `ConstrainedMCObjective`, each setup according to whether you want a given output larger or smaller than some bound (with negativity implying feasibility for each). Under the hood, the product of the sigmoid approximation of all constraints will be applied. Note that using too many constraints may end up smoothing out the objective a bit too much around the constraint boundaries and may make it harder to optimize if the optimal solutions are near the constraint boundaries.

**Comment by Balandat** [comment]
cc @SebastianAment re the constraints here.

**Comment by githubwl2021** [comment]
@saitcakmak Thanks a lot.
